File: utils/response.go
package utils

import "github.com/gofiber/fiber/v2"

type Pagination struct {
	Page     int `json:"page"`
	PageSize int `json:"page_size"`
	Total    int64 `json:"total"`
	Pages    int `json:"pages"`
}

type Response struct {
	Success    bool        `json:"success"`
	Data       interface{} `json:"data,omitempty"`
	Message    string      `json:"message,omitempty"`
	Pagination *Pagination `json:"pagination,omitempty"`
}

func SuccessResponse(c *fiber.Ctx, data interface{}) error {
	return c.JSON(Response{
		Success: true,
		Data:    data,
	})
}

func SuccessResponseWithPagination(c *fiber.Ctx, data interface{}, page, pageSize int, total int64) error {
	pages := int(total) / pageSize
	if int(total)%pageSize > 0 {
		pages++
	}

	return c.JSON(Response{
		Success: true,
		Data:    data,
		Pagination: &Pagination{
			Page:     page,
			PageSize: pageSize,
			Total:    total,
			Pages:    pages,
		},
	})
}

func ErrorResponse(c *fiber.Ctx, status int, message string) error {
	return c.Status(status).JSON(Response{
		Success: false,
		Message: message,
	})
}

File: server/utils/response.ts
import { v4 as uuidv4 } from 'uuid';
import { Response } from 'express';
import { ApiResponse, Pagination, ErrorCodes, BusinessError } from '../types';

export function success<T>(data: T, message = 'success', pagination?: Pagination): ApiResponse<T> {
  return {
    code: ErrorCodes.SUCCESS,
    message,
    data,
    timestamp: Date.now(),
    requestId: uuidv4(),
    pagination,
  };
}

export function error(code: number, message: string, details?: unknown): ApiResponse<null> {
  return {
    code,
    message,
    data: null,
    timestamp: Date.now(),
    requestId: uuidv4(),
  };
}

export function sendSuccess<T>(res: Response, data: T, message = 'success', pagination?: Pagination): Response {
  return res.json(success(data, message, pagination));
}

export function sendError(res: Response, err: unknown): Response {
  if (err instanceof BusinessError) {
    return res.status(err.code >= 50000 ? 500 : 400).json(
      error(err.code, err.message, err.details)
    );
  }
  
  if (err instanceof Error) {
    console.error('[Unhandled Error]', err);
    return res.status(500).json(
      error(ErrorCodes.BUSINESS_ERROR, err.message)
    );
  }
  
  return res.status(500).json(
    error(ErrorCodes.BUSINESS_ERROR, 'Internal Server Error')
  );
}

export function calculatePagination(total: number, page: number, pageSize: number): Pagination {
  return {
    page,
    pageSize,
    total,
    totalPages: Math.ceil(total / pageSize),
  };
}
